### ISSUE TITLE
Recommend that servers follow the steps of CORS preflight for easier troubleshooting

### ISSUE THREAD

**Issue by jub0bs** [whatwg/fetch#1588]
## TL;DR

[Section 3.2.3](https://fetch.spec.whatwg.org/#http-responses) contains the following passage:

> A successful HTTP response, i.e., one where the server developer intends to share it, to a CORS request can use any status, **as long as it includes the headers stated above with values matching up with the request**.
> 
> A successful HTTP response to a CORS-preflight request is similar, except it is restricted to an ok status, e.g., 200 or 204.
>
> Any other kind of HTTP response is not successful and will either end up not being shared or fail the CORS-preflight request. **Be aware that any work the server performs might nonetheless leak through side channels, such as timing.** If server developers wish to denote this explicitly, the 403 status can be used, **coupled with omitting the relevant headers**.

(my emphases)

I think this passage would benefit from some clarification. In particular, what useful information a side-channel attacker could glean from the presence of CORS headers in the preflight response is unclear to me. Also, what "this" means in

> If server developers wish to denote this explicitly

is unclear to me. I would also like to know the kind of timing side-channel attacks we should be worried about.

## More context

As [discussed](https://matrix.to/#/!AGetWbsMpFPdSgUrbs:matrix.org/$OgQ-Mt3qtH6ec-UsIsvNITsVJWuuU-ZQrnvuqhGS0DQ?via=matrix.org&via=mozilla.org&via=igalia.com) with @annevk on https://whatwg.org/chat, clarification of this passage may free developers of CORS middleware libraries to return more CORS headers than they typically do, which would make CORS issues easier to debug. Here is one example.

Suppose Alice configures CORS on `https://alice.com` to allow `https://alice-client.com` with request headers `Authorization`. However, in her client (running on `https://alice-client.com`), she also sends some `Foo` request header:

```javascript
fetch('//alice.com', {headers: {'Foo': 'Foo'}})
```

Obviously, CORS preflight then fails because the `Foo` header isn't allowed in Alice's CORS config. But most middleware libraries would omit the `Access-Control-Allow-Origin` header altogether from the response. When debugging the CORS error in the browser, Alice would be confronted with an error message like

> Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource.

which is correct but a bit misleading as to the actual reason why the server didn't respond with the required CORS headers (i.e. that request header `Foo` is not allowed).

In my experience, many CORS-related questions asked on Stack Overflow stem from this design decision. People are puzzled because they get such a misleading error message even when they do allow the client's origin in their CORS configuration.

However, this lack of debuggability can be remedied. Browser vendors could produce more verbose/informative messages, but that's probably quite a tall order. Instead, a CORS middleware library could be designed to respond to Alice's preflight request with

```http
Access-Control-Allow-Origin: https://alice-client.com
```
or perhaps with

```http
Access-Control-Allow-Origin: https://alice-client.com
Access-Control-Allow-Headers: authorization
```

Either way, the browser's resulting error message would be much more useful for debugging purposes:

> Request header field Foo is not allowed by Access-Control-Allow-Headers in preflight response.

Alice would then be in a better position to realise that she needs to either drop header `Foo` from her request or allow request-header field `Foo` in her CORS configuration.

**Comment by jub0bs** [comment]
@annevk I've reframed/retitled this issue. Normative requirements for servers that they follow the steps of CORS-preflight fetch would make CORS issues easier to debug and, thereby, alleviate the trouble of users who depend on CORS middleware libraries. More details at https://jub0bs.com/posts/2023-02-08-fearless-cors/#9-ease-troubleshooting-by-eschewing-shortcuts-during-preflight

**Comment by annevk** [comment]
The tradeoff here is that you end up giving more information to potential attackers. What is best in any given situation is not something we can make a universal statement about I think.

With regards to the timing channel, that is mainly for non-preflight responses, though perhaps attackers can also learn something from the amount of time it takes to grant or deny access to a particular request.

**Comment by jub0bs** [comment]
> The tradeoff here is that you end up giving more information to potential attackers.

That's true, but I wonder two things:

1. Would such information be valuable to an attacker?
2. Could an attacker not glean this information through dynamic analysis of the CORS-aware resource anyway?

In fact, if we assume that both 1. and 2. are true, typical CORS implementations (as described in [the OP](https://github.com/whatwg/fetch/issues/1588#issue-1526130185)) do encourage attackers to perform such dynamic analysis, which may place undue load on the server.

> [...] perhaps attackers can also learn something from the amount of time it takes to grant or deny access to a particular request.

Just to be clear, are you referring here purely to the _CORS-specific part_ of handling the _actual_ request, i.e. the logic in charge of checking whether the request's origin is allowed? If that's what you referring to, your warning in this regard still stands:

> Be aware that any work the server performs might nonetheless leak through side channels, such as timing.

But I don't see why this consideration should prevent us from pursuing the idea that servers follow the steps of the CORS-preflight fetch algorithm, so as to ease troubleshooting of CORS issues on the client side.

Perhaps a concrete/realistic attack scenario (it one exists) would clarify the issue. Can you think of one?

**Comment by annevk** [comment]
They are separate things, no?

1 is about how much information you expose directly, not even considering timing attacks. Perhaps dynamic analysis gives away some of that too, but maybe you'd get blocked for doing that. It depends.

2 is about how quickly you respond to any particular request. Perhaps the server can always respond in a certain amount of time, whether it accepts or not, avoiding leaking information this way.

**Comment by jub0bs** [comment]
1 My main question is whether the information thus revealed (not even considering timing attacks indeed) would be pertinent to an attacker. Only if the answer is "yes" should we worry about dynamic analysis getting detected/blocked, don't you think?

2 Do you have specific example in mind that would clearly weigh against the stronger normative server requirements that I suggested above?

Besides, you have to balance the ramifications of leaking such information against ease of troubleshooting. How many CORS misconfigurations with serious security consequences can be ascribed to the difficulty in troubleshooting CORS issues on the client side, due to CORS middleware hiding precious context from the browser?

In my experience, server developers on a deadline and at a loss to resolve a hard-to-debug CORS issue will sooner void the SOP's restrictions on network access (e.g. by allowing access from all origins with credentials) than get to the root cause of their CORS issue.

**Comment by jub0bs** [comment]
@sideshowbarker I'd be curious to hear your thoughts about this. 🙏

**Comment by sideshowbarker** [comment]
I haven’t been able to make time to consider this issue in detail yet, but from my experience over years of reading and answering CORS questions on Stack Overflow, I can say: It’s very clear that debugging CORS issues is a major pain point for web developers that costs developers a huge amount of time — and something they chronically make the same kinds of mistakes with. And I see nothing to indicate that situation has gotten any better over the years, or that it ever will.

Given that, I think we should do whatever we can safely to give web developers any addition debugging help that we can. And giving middleware authors some focused guidance in the spec about sending the `Access-Control-Allow-Origin` header under the circumstances outlined in the issue description seems to me at least like one specific change that could really help a lot of developers in practice.

**Comment by jub0bs** [comment]
Regardless of any changes to the Fetch standard, perhaps the best of both worlds is possible after all. FWIW, I'm working on a debug mode for my CORS middleware, which can be toggled (in a concurrency-safe manner) even as the middleware is processing requests.

The debug mode changes the behaviour of the middleware in the face of preflight failure:

- When debug is on, the middleware includes enough information in the preflight response for the browser to produce a helpful CORS error message; that way, server developers can easily troubleshoot their CORS issues by inspecting the browser's DevTools.
- When debug is off, the middleware omits all CORS headers from the preflight response, revealing nothing about CORS configuration.

Thanks again for your insight, @annevk and @sideshowbarker.

I think I'm happy enough with the approach outlined above to close this issue.
